# Pandas Series and DataFrame

In [1]:
import numpy as np 
import pandas as pd 

In [2]:
# help(pd.Series) 

myindex = ['USA', 'Canada', 'Mexico']
mydata  = [1776, 1867, 1821]

myser = pd.Series(data=mydata, index=myindex)
myser

USA       1776
Canada    1867
Mexico    1821
dtype: int64

In [3]:
myser['USA']

1776

In [4]:
ages = {
    'sam':10,
    'frank':20, 
    'david':30
}

In [5]:
pd.Series(ages)

sam      10
frank    20
david    30
dtype: int64

In [6]:
q1 = {'Japan': 80, 'China':450, 'India': 200, 'USA': 250}
q2 = {'Brazil': 100, 'China': 500, 'India': 210, 'USA': 260}

In [7]:
sales_q1 = pd.Series(q1)
sales_q2 = pd.Series(q2)

In [8]:
sales_q1

Japan     80
China    450
India    200
USA      250
dtype: int64

In [9]:
sales_q2

Brazil    100
China     500
India     210
USA       260
dtype: int64

In [10]:
sales_q1['Japan']

80

In [11]:
sales_q2['Brazil']

100

In [12]:
sales_q1.keys()

Index(['Japan', 'China', 'India', 'USA'], dtype='object')

In [13]:
sales_q1 * 2

Japan    160
China    900
India    400
USA      500
dtype: int64

In [14]:
sales_q1 + sales_q2

Brazil      NaN
China     950.0
India     410.0
Japan       NaN
USA       510.0
dtype: float64

In [15]:
sales_q1.add(sales_q2, fill_value=0)

Brazil    100.0
China     950.0
India     410.0
Japan      80.0
USA       510.0
dtype: float64

In [16]:
np.random.seed(101)
my_data = np.random.randint(0, 101, size=(4,3))

In [17]:
my_data

array([[95, 11, 81],
       [70, 63, 87],
       [75,  9, 77],
       [40,  4, 63]])

In [18]:
my_index = ['CA', 'NY', 'AZ', 'TX']
my_columns = ['Jan', 'Feb', 'Mar']

In [19]:
df = pd.DataFrame(my_data, index=my_index, columns=my_columns)
df

Jan  Feb  Mar
CA   95   11   81
NY   70   63   87
AZ   75    9   77
TX   40    4   63

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, CA to TX
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Jan     4 non-null      int64
 1   Feb     4 non-null      int64
 2   Mar     4 non-null      int64
dtypes: int64(3)
memory usage: 128.0+ bytes


In [21]:
ls

exported_car_sales  Pandas_2.ipynb  pandas-exercises.ipynb  Pandas.ipynb


In [22]:
df = pd.read_csv("https://docs.google.com/spreadsheets/d/1SAiQD2df3tt7JpAcfE_HV-DknR5Byp-EacIuIMUNnVM/export?gid=674589216&format=csv")
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name     CC Number Payment ID  
0    Christy Cunningham  3.560330e+15    Sun2959  
1        Douglas Tucker  4.478070e+15    Sun4608  
2        Travis Walters  6.011812e+15    Sun4458  
3      Nathaniel Harris  4.676138e+15    Sun5260  
4          Tonya Carter  4.832733e+15    Sun2251  
..                  ...           ...        ...  
239       Michael Avila  5.296069e+15    Sat2657  
240      Monica Sanders  3.506806e+15    Sat1766  
241          Keith Wong  6.011892e+15    Sat3880  
242        Dennis Dixon  4.375221e+12      Sat17  
243     Michelle Hardin  3.511452e+15    Thur672  

[244 rows x 11 columns]

In [23]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [24]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [25]:
df.head(10)

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   
5       25.29  4.71    Male     No  Sun  Dinner     4              6.32   
6        8.77  2.00    Male     No  Sun  Dinner     2              4.38   
7       26.88  3.12    Male     No  Sun  Dinner     4              6.72   
8       15.04  1.96    Male     No  Sun  Dinner     2              7.52   
9       14.78  3.23    Male     No  Sun  Dinner     2              7.39   

           Payer Name     CC Number Payment ID  
0  Christy Cunningham  3.560330e+15    Sun2959  
1      Douglas Tucker  4.478070e+15    Sun4608  
2      Travis Walters  6.011812e+15    Sun4458  
3    Nathaniel Harris  4.676138e+15    Sun5260  
4        Tonya Carter  4.832733e+15    Sun2251  
5          Erik Smith  2.131404e+14    Sun9679  
6  Kristopher Johnson  2.223728e+15    Sun5985  
7         Robert Buck  3.514785e+15    Sun8157  
8     Joseph Mcdonald  3.522866e+15    Sun6820  
9       Jerome Abbott  3.532125e+15    Sun3775

In [26]:
df.tail(10)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
234       15.53  3.00    Male    Yes   Sat  Dinner     2              7.76   
235       10.07  1.25    Male     No   Sat  Dinner     2              5.04   
236       12.60  1.00    Male    Yes   Sat  Dinner     2              6.30   
237       32.83  1.17    Male    Yes   Sat  Dinner     2             16.42   
238       35.83  4.67  Female     No   Sat  Dinner     3             11.94   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name     CC Number Payment ID  
234    Tracy Douglas  4.097940e+15    Sat7220  
235    Sean Gonzalez  3.534021e+15    Sat4615  
236    Matthew Myers  3.543676e+15    Sat5032  
237     Thomas Brown  4.284723e+15    Sat2929  
238   Kimberly Crane  6.761840e+11    Sat9777  
239    Michael Avila  5.296069e+15    Sat2657  
240   Monica Sanders  3.506806e+15    Sat1766  
241       Keith Wong  6.011892e+15    Sat3880  
242     Dennis Dixon  4.375221e+12      Sat17  
243  Michelle Hardin  3.511452e+15    Thur672

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    float64
 10  Payment ID        244 non-null    object 
dtypes: float64(4), int64(1), object(6)
memory usage: 21.1+ KB


In [28]:
df.describe()

total_bill         tip        size  price_per_person     CC Number
count  244.000000  244.000000  244.000000        244.000000  2.440000e+02
mean    19.785943    2.998279    2.569672          7.888197  2.563496e+15
std      8.902412    1.383638    0.951100          2.914234  2.369340e+15
min      3.070000    1.000000    1.000000          2.880000  6.040679e+10
25%     13.347500    2.000000    2.000000          5.800000  3.040731e+13
50%     17.795000    2.900000    2.000000          7.255000  3.525318e+15
75%     24.127500    3.562500    3.000000          9.390000  4.553675e+15
max     50.810000   10.000000    6.000000         20.270000  6.596454e+15

In [29]:
df.describe().transpose()

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.998279e+00  1.383638e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888197e+00  2.914234e+00  2.880000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.562500e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.800000e+00  7.255000e+00  9.390000e+00  2.027000e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15

In [30]:
df['total_bill']

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [31]:
df[['total_bill', 'tip', 'sex']]

total_bill   tip     sex
0         16.99  1.01  Female
1         10.34  1.66    Male
2         21.01  3.50    Male
3         23.68  3.31    Male
4         24.59  3.61  Female
..          ...   ...     ...
239       29.03  5.92    Male
240       27.18  2.00  Female
241       22.67  2.00    Male
242       17.82  1.75    Male
243       18.78  3.00  Female

[244 rows x 3 columns]

In [32]:
df['tip_percentage'] = 100 * (df['tip']/ df['total_bill'])
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name     CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3.560330e+15    Sun2959        5.944673  
1      Douglas Tucker  4.478070e+15    Sun4608       16.054159  
2      Travis Walters  6.011812e+15    Sun4458       16.658734  
3    Nathaniel Harris  4.676138e+15    Sun5260       13.978041  
4        Tonya Carter  4.832733e+15    Sun2251       14.680765

In [33]:
df['price_per_person'] = np.round(df['price_per_person'], 2)
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name     CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3.560330e+15    Sun2959        5.944673  
1      Douglas Tucker  4.478070e+15    Sun4608       16.054159  
2      Travis Walters  6.011812e+15    Sun4458       16.658734  
3    Nathaniel Harris  4.676138e+15    Sun5260       13.978041  
4        Tonya Carter  4.832733e+15    Sun2251       14.680765

In [34]:
df = df.drop(['tip_percentage'], axis=1)

In [35]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name     CC Number Payment ID  
0  Christy Cunningham  3.560330e+15    Sun2959  
1      Douglas Tucker  4.478070e+15    Sun4608  
2      Travis Walters  6.011812e+15    Sun4458  
3    Nathaniel Harris  4.676138e+15    Sun5260  
4        Tonya Carter  4.832733e+15    Sun2251

In [36]:
df  = df.set_index("Payment ID")
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name     CC Number  
Payment ID                                                      
Sun2959                 8.49  Christy Cunningham  3.560330e+15  
Sun4608                 3.45      Douglas Tucker  4.478070e+15  
Sun4458                 7.00      Travis Walters  6.011812e+15  
Sun5260                11.84    Nathaniel Harris  4.676138e+15  
Sun2251                 6.15        Tonya Carter  4.832733e+15

In [37]:
df = df.reset_index()
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name     CC Number  
0              8.49  Christy Cunningham  3.560330e+15  
1              3.45      Douglas Tucker  4.478070e+15  
2              7.00      Travis Walters  6.011812e+15  
3             11.84    Nathaniel Harris  4.676138e+15  
4              6.15        Tonya Carter  4.832733e+15

In [38]:
df  = df.set_index("Payment ID")
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name     CC Number  
Payment ID                                                      
Sun2959                 8.49  Christy Cunningham  3.560330e+15  
Sun4608                 3.45      Douglas Tucker  4.478070e+15  
Sun4458                 7.00      Travis Walters  6.011812e+15  
Sun5260                11.84    Nathaniel Harris  4.676138e+15  
Sun2251                 6.15        Tonya Carter  4.832733e+15

In [39]:
df.iloc[0]

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number           3560330000000000.0
Name: Sun2959, dtype: object

In [40]:
df.loc['Sun2959']

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number           3560330000000000.0
Name: Sun2959, dtype: object

In [41]:
df.iloc[0:4]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name     CC Number  
Payment ID                                                      
Sun2959                 8.49  Christy Cunningham  3.560330e+15  
Sun4608                 3.45      Douglas Tucker  4.478070e+15  
Sun4458                 7.00      Travis Walters  6.011812e+15  
Sun5260                11.84    Nathaniel Harris  4.676138e+15

In [42]:
df.loc[['Sun2959', 'Sun4608', 'Sun5260']]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name     CC Number  
Payment ID                                                      
Sun2959                 8.49  Christy Cunningham  3.560330e+15  
Sun4608                 3.45      Douglas Tucker  4.478070e+15  
Sun5260                11.84    Nathaniel Harris  4.676138e+15

In [43]:
df = df.drop('Sun2959', axis=0)
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No  Sun  Dinner     4   

            price_per_person        Payer Name     CC Number  
Payment ID                                                    
Sun4608                 3.45    Douglas Tucker  4.478070e+15  
Sun4458                 7.00    Travis Walters  6.011812e+15  
Sun5260                11.84  Nathaniel Harris  4.676138e+15  
Sun2251                 6.15      Tonya Carter  4.832733e+15  
Sun9679                 6.32        Erik Smith  2.131404e+14

In [44]:
row_one = df.iloc[0:1]
row_one

total_bill   tip   sex smoker  day    time  size  \
Payment ID                                                     
Sun4608          10.34  1.66  Male     No  Sun  Dinner     3   

            price_per_person      Payer Name     CC Number  
Payment ID                                                  
Sun4608                 3.45  Douglas Tucker  4.478070e+15

In [45]:
df = pd.concat([df, row_one])

In [46]:
df

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   

            price_per_person        Payer Name     CC Number  
Payment ID                                                    
Sun4608                 3.45    Douglas Tucker  4.478070e+15  
Sun4458                 7.00    Travis Walters  6.011812e+15  
Sun5260                11.84  Nathaniel Harris  4.676138e+15  
Sun2251                 6.15      Tonya Carter  4.832733e+15  
Sun9679                 6.32        Erik Smith  2.131404e+14  
...                      ...               ...           ...  
Sat1766                13.59    Monica Sanders  3.506806e+15  
Sat3880                11.34        Keith Wong  6.011892e+15  
Sat17                   8.91      Dennis Dixon  4.375221e+12  
Thur672                 9.39   Michelle Hardin  3.511452e+15  
Sun4608                 3.45    Douglas Tucker  4.478070e+15  

[244 rows x 10 columns]

In [47]:
df = pd.read_csv("https://docs.google.com/spreadsheets/d/1SAiQD2df3tt7JpAcfE_HV-DknR5Byp-EacIuIMUNnVM/export?gid=674589216&format=csv")
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name     CC Number Payment ID  
0  Christy Cunningham  3.560330e+15    Sun2959  
1      Douglas Tucker  4.478070e+15    Sun4608  
2      Travis Walters  6.011812e+15    Sun4458  
3    Nathaniel Harris  4.676138e+15    Sun5260  
4        Tonya Carter  4.832733e+15    Sun2251

In [48]:
df[df['total_bill'] > 40]

total_bill    tip     sex smoker   day    time  size  price_per_person  \
59        48.27   6.73    Male     No   Sat  Dinner     4             12.07   
95        40.17   4.73    Male    Yes   Fri  Dinner     4             10.04   
102       44.30   2.50  Female    Yes   Sat  Dinner     3             14.77   
142       41.19   5.00    Male     No  Thur   Lunch     5              8.24   
156       48.17   5.00    Male     No   Sun  Dinner     6              8.03   
170       50.81  10.00    Male    Yes   Sat  Dinner     3             16.94   
182       45.35   3.50    Male    Yes   Sun  Dinner     3             15.12   
184       40.55   3.00    Male    Yes   Sun  Dinner     2             20.27   
197       43.11   5.00  Female    Yes  Thur   Lunch     4             10.78   
212       48.33   9.00    Male     No   Sat  Dinner     4             12.08   

          Payer Name     CC Number Payment ID  
59       Brian Ortiz  6.596454e+15    Sat8139  
95     Aaron Bentley  1.800266e+14    Fri9628  
102    Heather Cohen  3.797711e+14    Sat6240  
142     Eric Andrews  4.356532e+15   Thur3621  
156    Ryan Gonzales  3.523151e+15    Sun7518  
170    Gregory Clark  5.473851e+15    Sat1954  
182     Jose Parsons  4.112210e+15    Sun2337  
184      Stephen Cox  3.547798e+15    Sun5140  
197      Brooke Soto  5.544902e+15   Thur9313  
212  Alex Williamson  6.762188e+11    Sat4590

In [49]:
df[df['sex'] == 'Male']

total_bill   tip   sex smoker  day    time  size  price_per_person  \
1         10.34  1.66  Male     No  Sun  Dinner     3              3.45   
2         21.01  3.50  Male     No  Sun  Dinner     3              7.00   
3         23.68  3.31  Male     No  Sun  Dinner     2             11.84   
5         25.29  4.71  Male     No  Sun  Dinner     4              6.32   
6          8.77  2.00  Male     No  Sun  Dinner     2              4.38   
..          ...   ...   ...    ...  ...     ...   ...               ...   
236       12.60  1.00  Male    Yes  Sat  Dinner     2              6.30   
237       32.83  1.17  Male    Yes  Sat  Dinner     2             16.42   
239       29.03  5.92  Male     No  Sat  Dinner     3              9.68   
241       22.67  2.00  Male    Yes  Sat  Dinner     2             11.34   
242       17.82  1.75  Male     No  Sat  Dinner     2              8.91   

             Payer Name     CC Number Payment ID  
1        Douglas Tucker  4.478070e+15    Sun4608  
2        Travis Walters  6.011812e+15    Sun4458  
3      Nathaniel Harris  4.676138e+15    Sun5260  
5            Erik Smith  2.131404e+14    Sun9679  
6    Kristopher Johnson  2.223728e+15    Sun5985  
..                  ...           ...        ...  
236       Matthew Myers  3.543676e+15    Sat5032  
237        Thomas Brown  4.284723e+15    Sat2929  
239       Michael Avila  5.296069e+15    Sat2657  
241          Keith Wong  6.011892e+15    Sat3880  
242        Dennis Dixon  4.375221e+12      Sat17  

[157 rows x 11 columns]

In [50]:
df[df['size']>3]

total_bill   tip     sex smoker   day    time  size  price_per_person  \
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
5         25.29  4.71    Male     No   Sun  Dinner     4              6.32   
7         26.88  3.12    Male     No   Sun  Dinner     4              6.72   
11        35.26  5.00  Female     No   Sun  Dinner     4              8.82   
13        18.43  3.00    Male     No   Sun  Dinner     4              4.61   
23        39.42  7.58    Male     No   Sat  Dinner     4              9.86   
25        17.81  2.34    Male     No   Sat  Dinner     4              4.45   
31        18.35  2.50    Male     No   Sat  Dinner     4              4.59   
33        20.69  2.45  Female     No   Sat  Dinner     4              5.17   
44        30.40  5.60    Male     No   Sun  Dinner     4              7.60   
47        32.40  6.00    Male     No   Sun  Dinner     4              8.10   
52        34.81  5.20  Female     No   Sun  Dinner     4              8.70   
54        25.56  4.34    Male     No   Sun  Dinner     4              6.39   
56        38.01  3.00    Male    Yes   Sat  Dinner     4              9.50   
59        48.27  6.73    Male     No   Sat  Dinner     4             12.07   
63        18.29  3.76    Male    Yes   Sat  Dinner     4              4.57   
77        27.20  4.00    Male     No  Thur   Lunch     4              6.80   
85        34.83  5.17  Female     No  Thur   Lunch     4              8.71   
95        40.17  4.73    Male    Yes   Fri  Dinner     4             10.04   
116       29.93  5.07    Male     No   Sun  Dinner     4              7.48   
119       24.08  2.92  Female     No  Thur   Lunch     4              6.02   
125       29.80  4.20  Female     No  Thur   Lunch     6              4.97   
141       34.30  6.70    Male     No  Thur   Lunch     6              5.72   
142       41.19  5.00    Male     No  Thur   Lunch     5              8.24   
143       27.05  5.00  Female     No  Thur   Lunch     6              4.51   
153       24.55  2.00    Male     No   Sun  Dinner     4              6.14   
154       19.77  2.00    Male     No   Sun  Dinner     4              4.94   
155       29.85  5.14  Female     No   Sun  Dinner     5              5.97   
156       48.17  5.00    Male     No   Sun  Dinner     6              8.03   
157       25.00  3.75  Female     No   Sun  Dinner     4              6.25   
159       16.49  2.00    Male     No   Sun  Dinner     4              4.12   
160       21.50  3.50    Male     No   Sun  Dinner     4              5.38   
167       31.71  4.50    Male     No   Sun  Dinner     4              7.93   
180       34.65  3.68    Male    Yes   Sun  Dinner     4              8.66   
183       23.17  6.50    Male    Yes   Sun  Dinner     4              5.79   
185       20.69  5.00    Male     No   Sun  Dinner     5              4.14   
187       30.46  2.00    Male    Yes   Sun  Dinner     5              6.09   
197       43.11  5.00  Female    Yes  Thur   Lunch     4             10.78   
204       20.53  4.00    Male    Yes  Thur   Lunch     4              5.13   
207       38.73  3.00    Male    Yes   Sat  Dinner     4              9.68   
211       25.89  5.16    Male    Yes   Sat  Dinner     4              6.47   
212       48.33  9.00    Male     No   Sat  Dinner     4             12.08   
216       28.15  3.00    Male    Yes   Sat  Dinner     5              5.63   
219       30.14  3.09  Female    Yes   Sat  Dinner     4              7.54   
227       20.45  3.00    Male     No   Sat  Dinner     4              5.11   
230       24.01  2.00    Male    Yes   Sat  Dinner     4              6.00   

                Payer Name     CC Number Payment ID  
4             Tonya Carter  4.832733e+15    Sun2251  
5               Erik Smith  2.131404e+14    Sun9679  
7              Robert Buck  3.514785e+15    Sun8157  
11            Diane Macias  4.577817e+15    Sun6686  
13            Joshua Jones  6.011160e+15    Sun2971  
23          La

In [51]:
df[(df['total_bill'] > 30) & (df['sex'] == 'Male')]

total_bill    tip   sex smoker   day    time  size  price_per_person  \
23        39.42   7.58  Male     No   Sat  Dinner     4              9.86   
39        31.27   5.00  Male     No   Sat  Dinner     3             10.42   
44        30.40   5.60  Male     No   Sun  Dinner     4              7.60   
47        32.40   6.00  Male     No   Sun  Dinner     4              8.10   
56        38.01   3.00  Male    Yes   Sat  Dinner     4              9.50   
59        48.27   6.73  Male     No   Sat  Dinner     4             12.07   
83        32.68   5.00  Male    Yes  Thur   Lunch     2             16.34   
95        40.17   4.73  Male    Yes   Fri  Dinner     4             10.04   
112       38.07   4.00  Male     No   Sun  Dinner     3             12.69   
141       34.30   6.70  Male     No  Thur   Lunch     6              5.72   
142       41.19   5.00  Male     No  Thur   Lunch     5              8.24   
156       48.17   5.00  Male     No   Sun  Dinner     6              8.03   
167       31.71   4.50  Male     No   Sun  Dinner     4              7.93   
170       50.81  10.00  Male    Yes   Sat  Dinner     3             16.94   
173       31.85   3.18  Male    Yes   Sun  Dinner     2             15.92   
175       32.90   3.11  Male    Yes   Sun  Dinner     2             16.45   
179       34.63   3.55  Male    Yes   Sun  Dinner     2             17.32   
180       34.65   3.68  Male    Yes   Sun  Dinner     4              8.66   
182       45.35   3.50  Male    Yes   Sun  Dinner     3             15.12   
184       40.55   3.00  Male    Yes   Sun  Dinner     2             20.27   
187       30.46   2.00  Male    Yes   Sun  Dinner     5              6.09   
207       38.73   3.00  Male    Yes   Sat  Dinner     4              9.68   
210       30.06   2.00  Male    Yes   Sat  Dinner     3             10.02   
212       48.33   9.00  Male     No   Sat  Dinner     4             12.08   
237       32.83   1.17  Male    Yes   Sat  Dinner     2             16.42   

                Payer Name     CC Number Payment ID  
23          Lance Peterson  3.542584e+15     Sat239  
39       Mr. Brandon Berry  6.011526e+15    Sat6373  
44             Todd Cooper  5.038468e+11    Sun2274  
47            James Barnes  3.552003e+15    Sun9677  
56   James Christensen DDS  3.497936e+14    Sat8903  
59             Brian Ortiz  6.596454e+15    Sat8139  
83           Daniel Murphy  5.356178e+15   Thur8801  
95           Aaron Bentley  1.800266e+14    Fri9628  
112             Jeff Lopez  3.572866e+15     Sun591  
141         Steven Carlson  3.526516e+15   Thur1025  
142           Eric Andrews  4.356532e+15   Thur3621  
156          Ryan Gonzales  3.523151e+15    Sun7518  
167         Michael Lawson  3.566286e+15    Sun3719  
170          Gregory Clark  5.473851e+15    Sat1954  
173            Scott Perez  3.577116e+15    Sun9335  
175        Nathan Reynolds  3.703070e+14    Sun5109  
179           Brian Bailey  3.466563e+14    Sun9851  
180       James Hebert DDS  6.761687e+11    Sun7544  
182           Jose Parsons  4.112210e+15    Sun2337  
184            Stephen Cox  3.547798e+15    Sun5140  
187          David Barrett  4.792883e+15    Sun9987  
207          Ricky Ramirez  3.478180e+14    Sat4505  
210          Shawn Mendoza  3.018405e+13    Sat8361  
212        Alex Williamson  6.762188e+11    Sat4590  
237           Thomas Brown  4.284723e+15    Sat2929

In [52]:
df[(df['day'] == 'Sat') | (df['day'] == 'Sun') | (df['day'] == 'Fri')]

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0         16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No  Sun  Dinner     4              6.15   
..          ...   ...     ...    ...  ...     ...   ...               ...   
238       35.83  4.67  Female     No  Sat  Dinner     3             11.94   
239       29.03  5.92    Male     No  Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes  Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes  Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No  Sat  Dinner     2              8.91   

             Payer Name     CC Number Payment ID  
0    Christy Cunningham  3.560330e+15    Sun2959  
1        Douglas Tucker  4.478070e+15    Sun4608  
2        Travis Walters  6.011812e+15    Sun4458  
3      Nathaniel Harris  4.676138e+15    Sun5260  
4          Tonya Carter  4.832733e+15    Sun2251  
..                  ...           ...        ...  
238      Kimberly Crane  6.761840e+11    Sat9777  
239       Michael Avila  5.296069e+15    Sat2657  
240      Monica Sanders  3.506806e+15    Sat1766  
241          Keith Wong  6.011892e+15    Sat3880  
242        Dennis Dixon  4.375221e+12      Sat17  

[182 rows x 11 columns]

In [53]:
df[df['day'].isin(['Sat', 'Sun'])]

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0         16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No  Sun  Dinner     4              6.15   
..          ...   ...     ...    ...  ...     ...   ...               ...   
238       35.83  4.67  Female     No  Sat  Dinner     3             11.94   
239       29.03  5.92    Male     No  Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes  Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes  Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No  Sat  Dinner     2              8.91   

             Payer Name     CC Number Payment ID  
0    Christy Cunningham  3.560330e+15    Sun2959  
1        Douglas Tucker  4.478070e+15    Sun4608  
2        Travis Walters  6.011812e+15    Sun4458  
3      Nathaniel Harris  4.676138e+15    Sun5260  
4          Tonya Carter  4.832733e+15    Sun2251  
..                  ...           ...        ...  
238      Kimberly Crane  6.761840e+11    Sat9777  
239       Michael Avila  5.296069e+15    Sat2657  
240      Monica Sanders  3.506806e+15    Sat1766  
241          Keith Wong  6.011892e+15    Sat3880  
242        Dennis Dixon  4.375221e+12      Sat17  

[163 rows x 11 columns]

In [54]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name     CC Number Payment ID  
0  Christy Cunningham  3.560330e+15    Sun2959  
1      Douglas Tucker  4.478070e+15    Sun4608  
2      Travis Walters  6.011812e+15    Sun4458  
3    Nathaniel Harris  4.676138e+15    Sun5260  
4        Tonya Carter  4.832733e+15    Sun2251

In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    float64
 10  Payment ID        244 non-null    object 
dtypes: float64(4), int64(1), object(6)
memory usage: 21.1+ KB


In [56]:
def last_num(num):
    return str(num)[-4:]
    
df['CC Number'] = df['CC Number'].apply(int)
df['last_num'] = df['CC Number'].apply(last_num)
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID last_num  
0  Christy Cunningham  3560330000000000    Sun2959     0000  
1      Douglas Tucker  4478070000000000    Sun4608     0000  
2      Travis Walters  6011812112971322    Sun4458     1322  
3    Nathaniel Harris  4676137647685994    Sun5260     5994  
4        Tonya Carter  4832732618637221    Sun2251     7221

In [57]:
df['total_bill'].mean()

19.78594262295082

In [58]:
def yelp(price):
    if price < 10:
        return "$"
    elif price >= 10 and price < 30:
        return "$$"
    else:
        return "$$$"

df['yelp'] = df['total_bill'].apply(yelp)
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID last_num yelp  
0  Christy Cunningham  3560330000000000    Sun2959     0000   $$  
1      Douglas Tucker  4478070000000000    Sun4608     0000   $$  
2      Travis Walters  6011812112971322    Sun4458     1322   $$  
3    Nathaniel Harris  4676137647685994    Sun5260     5994   $$  
4        Tonya Carter  4832732618637221    Sun2251     7221   $$

In [59]:
df['total_bill'].apply(lambda num: num*2)

0      33.98
1      20.68
2      42.02
3      47.36
4      49.18
       ...  
239    58.06
240    54.36
241    45.34
242    35.64
243    37.56
Name: total_bill, Length: 244, dtype: float64

In [60]:
def quality(total_bill, tip):
    if tip/total_bill > 0.25:
        return "Generous"
    else:
        return "Others"

df['Quality'] = df[['total_bill', 'tip']].apply(lambda df: quality(df['total_bill'], df['tip']), axis=1)

In [61]:
df['Quality'] = np.vectorize(quality)(df['total_bill'], df['tip'])

In [62]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID last_num yelp Quality  
0  Christy Cunningham  3560330000000000    Sun2959     0000   $$  Others  
1      Douglas Tucker  4478070000000000    Sun4608     0000   $$  Others  
2      Travis Walters  6011812112971322    Sun4458     1322   $$  Others  
3    Nathaniel Harris  4676137647685994    Sun5260     5994   $$  Others  
4        Tonya Carter  4832732618637221    Sun2251     7221   $$  Others

In [63]:
df = pd.read_csv("https://docs.google.com/spreadsheets/d/1SAiQD2df3tt7JpAcfE_HV-DknR5Byp-EacIuIMUNnVM/export?gid=674589216&format=csv")
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name     CC Number Payment ID  
0  Christy Cunningham  3.560330e+15    Sun2959  
1      Douglas Tucker  4.478070e+15    Sun4608  
2      Travis Walters  6.011812e+15    Sun4458  
3    Nathaniel Harris  4.676138e+15    Sun5260  
4        Tonya Carter  4.832733e+15    Sun2251

In [64]:
df.describe()

total_bill         tip        size  price_per_person     CC Number
count  244.000000  244.000000  244.000000        244.000000  2.440000e+02
mean    19.785943    2.998279    2.569672          7.888197  2.563496e+15
std      8.902412    1.383638    0.951100          2.914234  2.369340e+15
min      3.070000    1.000000    1.000000          2.880000  6.040679e+10
25%     13.347500    2.000000    2.000000          5.800000  3.040731e+13
50%     17.795000    2.900000    2.000000          7.255000  3.525318e+15
75%     24.127500    3.562500    3.000000          9.390000  4.553675e+15
max     50.810000   10.000000    6.000000         20.270000  6.596454e+15

In [65]:
df.sort_values('tip', ascending=False)

total_bill    tip     sex smoker   day    time  size  price_per_person  \
170       50.81  10.00    Male    Yes   Sat  Dinner     3             16.94   
212       48.33   9.00    Male     No   Sat  Dinner     4             12.08   
23        39.42   7.58    Male     No   Sat  Dinner     4              9.86   
59        48.27   6.73    Male     No   Sat  Dinner     4             12.07   
141       34.30   6.70    Male     No  Thur   Lunch     6              5.72   
..          ...    ...     ...    ...   ...     ...   ...               ...   
0         16.99   1.01  Female     No   Sun  Dinner     2              8.49   
236       12.60   1.00    Male    Yes   Sat  Dinner     2              6.30   
111        7.25   1.00  Female     No   Sat  Dinner     1              7.25   
67         3.07   1.00  Female    Yes   Sat  Dinner     1              3.07   
92         5.75   1.00  Female    Yes   Fri  Dinner     2              2.88   

             Payer Name     CC Number Payment ID  
170       Gregory Clark  5.473851e+15    Sat1954  
212     Alex Williamson  6.762188e+11    Sat4590  
23       Lance Peterson  3.542584e+15     Sat239  
59          Brian Ortiz  6.596454e+15    Sat8139  
141      Steven Carlson  3.526516e+15   Thur1025  
..                  ...           ...        ...  
0    Christy Cunningham  3.560330e+15    Sun2959  
236       Matthew Myers  3.543676e+15    Sat5032  
111         Terri Jones  3.559221e+15    Sat4801  
67        Tiffany Brock  4.359489e+15    Sat3455  
92         Leah Ramirez  3.508912e+15    Fri3780  

[244 rows x 11 columns]

In [66]:
df['total_bill'].max(), df['total_bill'].idxmax()

(50.81, 170)

In [67]:
df['total_bill'].min(), df['total_bill'].idxmin()

(3.07, 67)

In [68]:
df[['total_bill'	,'tip',	'size',	'price_per_person',	'CC Number']].corr()

total_bill       tip      size  price_per_person  CC Number
total_bill          1.000000  0.675734  0.598315          0.647554   0.104576
tip                 0.675734  1.000000  0.489299          0.347405   0.110857
size                0.598315  0.489299  1.000000         -0.175359  -0.030239
price_per_person    0.647554  0.347405 -0.175359          1.000000   0.135240
CC Number           0.104576  0.110857 -0.030239          0.135240   1.000000

In [69]:
df['day'].value_counts()

day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64

In [70]:
df['day'].unique()

array(['Sun', 'Sat', 'Thur', 'Fri'], dtype=object)

In [71]:
df['day'].nunique()

4

In [72]:
df['sex'].replace(['Female', 'Male'], ['F', 'M'])

0      F
1      M
2      M
3      M
4      F
      ..
239    M
240    F
241    M
242    M
243    F
Name: sex, Length: 244, dtype: object

In [73]:
df['sex'].map({'Female': 'F', 'Male': 'M'})

0      F
1      M
2      M
3      M
4      F
      ..
239    M
240    F
241    M
242    M
243    F
Name: sex, Length: 244, dtype: object

In [74]:
df.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
239    False
240    False
241    False
242    False
243    False
Length: 244, dtype: bool

In [75]:
df[df['total_bill'].between(10, 20)]

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
8         15.04  1.96    Male     No   Sun  Dinner     2              7.52   
9         14.78  3.23    Male     No   Sun  Dinner     2              7.39   
10        10.27  1.71    Male     No   Sun  Dinner     2              5.14   
..          ...   ...     ...    ...   ...     ...   ...               ...   
234       15.53  3.00    Male    Yes   Sat  Dinner     2              7.76   
235       10.07  1.25    Male     No   Sat  Dinner     2              5.04   
236       12.60  1.00    Male    Yes   Sat  Dinner     2              6.30   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name     CC Number Payment ID  
0    Christy Cunningham  3.560330e+15    Sun2959  
1        Douglas Tucker  4.478070e+15    Sun4608  
8       Joseph Mcdonald  3.522866e+15    Sun6820  
9         Jerome Abbott  3.532125e+15    Sun3775  
10        William Riley  5.662876e+11    Sun2546  
..                  ...           ...        ...  
234       Tracy Douglas  4.097940e+15    Sat7220  
235       Sean Gonzalez  3.534021e+15    Sat4615  
236       Matthew Myers  3.543676e+15    Sat5032  
242        Dennis Dixon  4.375221e+12      Sat17  
243     Michelle Hardin  3.511452e+15    Thur672  

[130 rows x 11 columns]

In [76]:
df.nlargest(3, columns='tip')

total_bill    tip   sex smoker  day    time  size  price_per_person  \
170       50.81  10.00  Male    Yes  Sat  Dinner     3             16.94   
212       48.33   9.00  Male     No  Sat  Dinner     4             12.08   
23        39.42   7.58  Male     No  Sat  Dinner     4              9.86   

          Payer Name     CC Number Payment ID  
170    Gregory Clark  5.473851e+15    Sat1954  
212  Alex Williamson  6.762188e+11    Sat4590  
23    Lance Peterson  3.542584e+15     Sat239

In [77]:
df.nsmallest(3, columns='tip')

total_bill  tip     sex smoker  day    time  size  price_per_person  \
67         3.07  1.0  Female    Yes  Sat  Dinner     1              3.07   
92         5.75  1.0  Female    Yes  Fri  Dinner     2              2.88   
111        7.25  1.0  Female     No  Sat  Dinner     1              7.25   

        Payer Name     CC Number Payment ID  
67   Tiffany Brock  4.359489e+15    Sat3455  
92    Leah Ramirez  3.508912e+15    Fri3780  
111    Terri Jones  3.559221e+15    Sat4801

In [78]:
df.sample(frac=0.1)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
169       10.63  2.00  Female    Yes   Sat  Dinner     2              5.32   
161       12.66  2.50    Male     No   Sun  Dinner     2              6.33   
69        15.01  2.09    Male    Yes   Sat  Dinner     2              7.50   
222        8.58  1.92    Male    Yes   Fri   Lunch     1              8.58   
93        16.32  4.30  Female    Yes   Fri  Dinner     2              8.16   
229       22.12  2.88  Female    Yes   Sat  Dinner     2             11.06   
56        38.01  3.00    Male    Yes   Sat  Dinner     4              9.50   
166       20.76  2.24    Male     No   Sun  Dinner     2             10.38   
172        7.25  5.15    Male    Yes   Sun  Dinner     2              3.62   
177       14.48  2.00    Male    Yes   Sun  Dinner     2              7.24   
185       20.69  5.00    Male     No   Sun  Dinner     5              4.14   
135        8.51  1.25  Female     No  Thur   Lunch     2              4.26   
33        20.69  2.45  Female     No   Sat  Dinner     4              5.17   
215       12.90  1.10  Female    Yes   Sat  Dinner     2              6.45   
85        34.83  5.17  Female     No  Thur   Lunch     4              8.71   
188       18.15  3.50  Female    Yes   Sun  Dinner     3              6.05   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   
48        28.55  2.05    Male     No   Sun  Dinner     3              9.52   
95        40.17  4.73    Male    Yes   Fri  Dinner     4             10.04   
183       23.17  6.50    Male    Yes   Sun  Dinner     4              5.79   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
75        10.51  1.25    Male     No   Sat  Dinner     2              5.26   
32        15.06  3.00  Female     No   Sat  Dinner     2              7.53   
210       30.06  2.00    Male    Yes   Sat  Dinner     3             10.02   

                Payer Name     CC Number Payment ID  
169               Amy Hill  3.536332e+15    Sat1788  
161        Brandon Oconnor  4.406882e+15    Sun5879  
69               Adam Hall  4.700924e+15     Sat855  
222         Jason Lawrence  3.505303e+15    Fri6624  
93          Natalie Nguyen  5.181236e+15    Fri6963  
229       Jennifer Russell  4.793003e+12    Sat3943  
56   James Christensen DDS  3.497936e+14    Sat8903  
166            Gordon Lane  4.110600e+15    Sun6738  
172            Larry White  3.043262e+13    Sun9209  
177            John Dudley  4.565183e+15    Sun6203  
185          Joseph Howell  3.036241e+13    Sun5842  
135         Rebecca Harris  4.320272e+15   Thur6600  
33           Amber Francis  3.777430e+14    Sat6649  
215           Jessica Owen  4.726905e+12    Sat6983  
85             Shawna Cook  6.011790e+15   Thur7972  
188         Glenda Wiggins  5.783293e+11     Sun430  
243        Michelle Hardin  3.511452e+15    Thur672  
48           Austin Fisher  6.011482e+15    Sun4142  
95           Aaron Bentley  1.800266e+14    Fri9628  
183      Dr. Michael James  4.718502e+12    Sun6059  
4             Tonya Carter  4.832733e+15    Sun2251  
75           Kenneth Hayes  2.131421e+14    Sat5056  
32           Amanda Wilson  2.131863e+14    Sat1327  
210          Shawn Mendoza  3.018405e+13    Sat8361

In [79]:
df = pd.read_csv("https://docs.google.com/spreadsheets/d/1jv3jA4KSvju8z89PmXk-j3WSECWMBeCJGXhsn867hck/export?gid=1682123716&format=csv")
df

first_name last_name   age  sex  pre_movie_score  post_movie_score
0        Tom     Hanks  63.0    m              8.0              10.0
1        NaN       NaN   NaN  NaN              NaN               NaN
2       Hugh   Jackman  51.0    m              NaN               NaN
3      Oprah   Winfrey  66.0    f              6.0               8.0
4       Emma     Stone  31.0    f              7.0               9.0

In [80]:
df.isnull()

first_name  last_name    age    sex  pre_movie_score  post_movie_score
0       False      False  False  False            False             False
1        True       True   True   True             True              True
2       False      False  False  False             True              True
3       False      False  False  False            False             False
4       False      False  False  False            False             False

In [81]:
df['pre_movie_score'].isnull()

0    False
1     True
2     True
3    False
4    False
Name: pre_movie_score, dtype: bool

In [82]:
df[df['pre_movie_score'].isnull() & df['sex'].notnull()]

first_name last_name   age sex  pre_movie_score  post_movie_score
2       Hugh   Jackman  51.0   m              NaN               NaN

In [83]:
df.dropna()

first_name last_name   age sex  pre_movie_score  post_movie_score
0        Tom     Hanks  63.0   m              8.0              10.0
3      Oprah   Winfrey  66.0   f              6.0               8.0
4       Emma     Stone  31.0   f              7.0               9.0

In [84]:
df.dropna(thresh=2)

first_name last_name   age sex  pre_movie_score  post_movie_score
0        Tom     Hanks  63.0   m              8.0              10.0
2       Hugh   Jackman  51.0   m              NaN               NaN
3      Oprah   Winfrey  66.0   f              6.0               8.0
4       Emma     Stone  31.0   f              7.0               9.0

In [85]:
df.dropna(axis=1, thresh=4)

first_name last_name   age  sex
0        Tom     Hanks  63.0    m
1        NaN       NaN   NaN  NaN
2       Hugh   Jackman  51.0    m
3      Oprah   Winfrey  66.0    f
4       Emma     Stone  31.0    f

In [86]:
df.dropna(subset=['last_name'])

first_name last_name   age sex  pre_movie_score  post_movie_score
0        Tom     Hanks  63.0   m              8.0              10.0
2       Hugh   Jackman  51.0   m              NaN               NaN
3      Oprah   Winfrey  66.0   f              6.0               8.0
4       Emma     Stone  31.0   f              7.0               9.0

In [87]:
df.fillna("New_value")

first_name  last_name        age        sex pre_movie_score post_movie_score
0        Tom      Hanks       63.0          m             8.0             10.0
1  New_value  New_value  New_value  New_value       New_value        New_value
2       Hugh    Jackman       51.0          m       New_value        New_value
3      Oprah    Winfrey       66.0          f             6.0              8.0
4       Emma      Stone       31.0          f             7.0              9.0

In [88]:
df['pre_movie_score'].fillna(0)

0    8.0
1    0.0
2    0.0
3    6.0
4    7.0
Name: pre_movie_score, dtype: float64

In [89]:
df['pre_movie_score'].fillna(df['pre_movie_score'].mean())

0    8.0
1    7.0
2    7.0
3    6.0
4    7.0
Name: pre_movie_score, dtype: float64

In [90]:
df['pre_movie_score'].interpolate()

0    8.000000
1    7.333333
2    6.666667
3    6.000000
4    7.000000
Name: pre_movie_score, dtype: float64

In [91]:
df = pd.read_csv("https://docs.google.com/spreadsheets/d/1-KOWPFut0N3LF3q1-O5o1MXe_ply4A9kR-SymTQvhFk/export?gid=1938795784&format=csv")
df

mpg  cylinders  displacement horsepower  weight  acceleration  \
0    18.0          8         307.0        130    3504          12.0   
1    15.0          8         350.0        165    3693          11.5   
2    18.0          8         318.0        150    3436          11.0   
3    16.0          8         304.0        150    3433          12.0   
4    17.0          8         302.0        140    3449          10.5   
..    ...        ...           ...        ...     ...           ...   
393  27.0          4         140.0         86    2790          15.6   
394  44.0          4          97.0         52    2130          24.6   
395  32.0          4         135.0         84    2295          11.6   
396  28.0          4         120.0         79    2625          18.6   
397  31.0          4         119.0         82    2720          19.4   

     model_year  origin                       name  
0            70       1  chevrolet chevelle malibu  
1            70       1          buick skylark 320  
2            70       1         plymouth satellite  
3            70       1              amc rebel sst  
4            70       1                ford torino  
..          ...     ...                        ...  
393          82       1            ford mustang gl  
394          82       2                  vw pickup  
395          82       1              dodge rampage  
396          82       1                ford ranger  
397          82       1                 chevy s-10  

[398 rows x 9 columns]

In [92]:
df['model_year'].value_counts()

model_year
73    40
78    36
76    34
82    31
75    30
70    29
79    29
80    29
81    29
71    28
72    28
77    28
74    27
Name: count, dtype: int64

In [93]:
df.groupby('model_year').mean(numeric_only=True)
# df.info()

mpg  cylinders  displacement       weight  acceleration  \
model_year                                                                  
70          17.689655   6.758621    281.413793  3372.793103     12.948276   
71          21.250000   5.571429    209.750000  2995.428571     15.142857   
72          18.714286   5.821429    218.375000  3237.714286     15.125000   
73          17.100000   6.375000    256.875000  3419.025000     14.312500   
74          22.703704   5.259259    171.740741  2877.925926     16.203704   
75          20.266667   5.600000    205.533333  3176.800000     16.050000   
76          21.573529   5.647059    197.794118  3078.735294     15.941176   
77          23.375000   5.464286    191.392857  2997.357143     15.435714   
78          24.061111   5.361111    177.805556  2861.805556     15.805556   
79          25.093103   5.827586    206.689655  3055.344828     15.813793   
80          33.696552   4.137931    115.827586  2436.655172     16.934483   
81          30.334483   4.620690    135.310345  2522.931034     16.306897   
82          31.709677   4.193548    128.870968  2453.548387     16.638710   

              origin  
model_year            
70          1.310345  
71          1.428571  
72          1.535714  
73          1.375000  
74          1.666667  
75          1.466667  
76          1.470588  
77          1.571429  
78          1.611111  
79          1.275862  
80          2.206897  
81          1.965517  
82          1.645161

In [94]:
df.groupby(['model_year', 'cylinders']).mean(numeric_only=True)

mpg  displacement       weight  acceleration  \
model_year cylinders                                                       
70         4          25.285714    107.000000  2292.571429     16.000000   
           6          20.500000    199.000000  2710.500000     15.500000   
           8          14.111111    367.555556  3940.055556     11.194444   
71         4          27.461538    101.846154  2056.384615     16.961538   
           6          18.000000    243.375000  3171.875000     14.750000   
           8          13.428571    371.714286  4537.714286     12.214286   
72         3          19.000000     70.000000  2330.000000     13.500000   
           4          23.428571    111.535714  2382.642857     17.214286   
           8          13.615385    344.846154  4228.384615     13.000000   
73         3          18.000000     70.000000  2124.000000     13.500000   
           4          22.727273    109.272727  2338.090909     17.136364   
           6          19.000000    212.250000  2917.125000     15.687500   
           8          13.200000    365.250000  4279.050000     12.250000   
74         4          27.800000     96.533333  2151.466667     16.400000   
           6          17.857143    230.428571  3320.000000     16.857143   
           8          14.200000    315.200000  4438.400000     14.700000   
75         4          25.250000    114.833333  2489.250000     15.833333   
           6          17.583333    233.750000  3398.333333     17.708333   
           8          15.666667    330.500000  4108.833333     13.166667   
76         4          26.766667    106.333333  2306.600000     16.866667   
           6          20.000000    221.400000  3349.600000     17.000000   
           8          14.666667    324.000000  4064.666667     13.222222   
77         3          21.500000     80.000000  2720.000000     13.500000   
           4          29.107143    106.500000  2205.071429     16.064286   
           6          19.500000    220.400000  3383.000000     16.900000   
           8          16.000000    335.750000  4177.500000     13.662500   
78         4          29.576471    112.117647  2296.764706     16.282353   
           5          20.300000    131.000000  2830.000000     15.900000   
           6          19.066667    213.250000  3314.166667     16.391667   
           8          19.050000    300.833333  3563.333333     13.266667   
79         4          31.525000    113.583333  2357.583333     15.991667   
           5          25.400000    183.000000  3530.000000     20.100000   
           6          22.950000    205.666667  3025.833333     15.433333   
           8          18.630000    321.400000  3862.900000     15.400000   
80         3          23.700000     70.000000  2420.000000     12.500000   
           4          34.612000    111.000000  2360.080000     17.144000   
           5          36.400000    121.000000  2950.000000     19.900000   
           6          25.900000    196.500000  3145.500000     15.050000   
81         4          32.814286    108.857143  2275.476190     16.466667   
           6          23.428571    184.000000  3093.571429     15.442857   
           8          26.600000    350.000000  3725.000000     19.000000   
82         4          32.071429    118.571429  2402.321429     16.703571   
           6          28.333333    225.000000  2931.666667     16.033333   

                        origin  
model_year cylinders            
70         4          2.285714  
           6          1.000000  
           8          1.000000  
71         4          1.923077  
           6          1.000000  
           8          1.000000  
72         3          3.000000  
           4          1.928571  
           8          1.000000  
73         3          3.000000  
           4          2.000000  
           6          1.250000  
           8          1.000000  
74         4          2.200000  
           6          1.000000  
           8          1.000000  
75         4 

In [95]:
df.groupby(['model_year', 'cylinders']).mean(numeric_only=True).index

MultiIndex([(70, 4),
            (70, 6),
            (70, 8),
            (71, 4),
            (71, 6),
            (71, 8),
            (72, 3),
            (72, 4),
            (72, 8),
            (73, 3),
            (73, 4),
            (73, 6),
            (73, 8),
            (74, 4),
            (74, 6),
            (74, 8),
            (75, 4),
            (75, 6),
            (75, 8),
            (76, 4),
            (76, 6),
            (76, 8),
            (77, 3),
            (77, 4),
            (77, 6),
            (77, 8),
            (78, 4),
            (78, 5),
            (78, 6),
            (78, 8),
            (79, 4),
            (79, 5),
            (79, 6),
            (79, 8),
            (80, 3),
            (80, 4),
            (80, 5),
            (80, 6),
            (81, 4),
            (81, 6),
            (81, 8),
            (82, 4),
            (82, 6)],
           names=['model_year', 'cylinders'])

In [96]:
year_cyl = df.groupby(['model_year', 'cylinders']).mean(numeric_only=True)

In [97]:
year_cyl.index.names

FrozenList(['model_year', 'cylinders'])

In [98]:
year_cyl.index.levels

FrozenList([[70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82], [3, 4, 5, 6, 8]])

In [99]:
year_cyl.loc[70]

mpg  displacement       weight  acceleration    origin
cylinders                                                              
4          25.285714    107.000000  2292.571429     16.000000  2.285714
6          20.500000    199.000000  2710.500000     15.500000  1.000000
8          14.111111    367.555556  3940.055556     11.194444  1.000000

In [100]:
year_cyl.loc[[70, 82]]

mpg  displacement       weight  acceleration  \
model_year cylinders                                                       
70         4          25.285714    107.000000  2292.571429     16.000000   
           6          20.500000    199.000000  2710.500000     15.500000   
           8          14.111111    367.555556  3940.055556     11.194444   
82         4          32.071429    118.571429  2402.321429     16.703571   
           6          28.333333    225.000000  2931.666667     16.033333   

                        origin  
model_year cylinders            
70         4          2.285714  
           6          1.000000  
           8          1.000000  
82         4          1.714286  
           6          1.000000

In [101]:
year_cyl.loc[(70,4)]

mpg               25.285714
displacement     107.000000
weight          2292.571429
acceleration      16.000000
origin             2.285714
Name: (70, 4), dtype: float64

In [102]:
year_cyl.xs(level='model_year', key=70)

mpg  displacement       weight  acceleration    origin
cylinders                                                              
4          25.285714    107.000000  2292.571429     16.000000  2.285714
6          20.500000    199.000000  2710.500000     15.500000  1.000000
8          14.111111    367.555556  3940.055556     11.194444  1.000000

In [103]:
year_cyl.xs(level='cylinders', key=4)

mpg  displacement       weight  acceleration    origin
model_year                                                              
70          25.285714    107.000000  2292.571429     16.000000  2.285714
71          27.461538    101.846154  2056.384615     16.961538  1.923077
72          23.428571    111.535714  2382.642857     17.214286  1.928571
73          22.727273    109.272727  2338.090909     17.136364  2.000000
74          27.800000     96.533333  2151.466667     16.400000  2.200000
75          25.250000    114.833333  2489.250000     15.833333  2.166667
76          26.766667    106.333333  2306.600000     16.866667  1.866667
77          29.107143    106.500000  2205.071429     16.064286  1.857143
78          29.576471    112.117647  2296.764706     16.282353  2.117647
79          31.525000    113.583333  2357.583333     15.991667  1.583333
80          34.612000    111.000000  2360.080000     17.144000  2.200000
81          32.814286    108.857143  2275.476190     16.466667  2.095238
82          32.071429    118.571429  2402.321429     16.703571  1.714286

In [104]:
numeric_df = df.select_dtypes(include=[np.number])
 
# Apply aggregation functions to the filtered DataFrame
numeric_df.agg(['std', 'mean'])

mpg  cylinders  displacement       weight  acceleration  \
std    7.815984   1.701004    104.269838   846.841774      2.757689   
mean  23.514573   5.454774    193.425879  2970.424623     15.568090   

      model_year    origin  
std     3.697627  0.802055  
mean   76.010050  1.572864

In [105]:
numeric_df.aggregate(['mean'])

mpg  cylinders  displacement       weight  acceleration  \
mean  23.514573   5.454774    193.425879  2970.424623      15.56809   

      model_year    origin  
mean    76.01005  1.572864

In [106]:
numeric_df.agg({'mpg':['max', 'mean'], 'weight':['mean', 'std']})

mpg       weight
max   46.600000          NaN
mean  23.514573  2970.424623
std         NaN   846.841774

In [107]:
data_one = pd.DataFrame(
    {
        "A": ["A0", "A1", "A2", "A3"],
        "B": ["B0", "B1", "B2", "B3"],
    },
    index=[0, 1, 2, 3],
)
data_one

A   B
0  A0  B0
1  A1  B1
2  A2  B2
3  A3  B3

In [108]:
data_two = pd.DataFrame(
    {
        "C": ["C0", "C1", "C2", "C3"],
        "D": ["D0", "D1", "D2", "D3"],
    },
    index=[0, 1, 2, 3],
)
data_two

C   D
0  C0  D0
1  C1  D1
2  C2  D2
3  C3  D3

In [109]:
pd.concat([data_one, data_two], axis=1)

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3

In [110]:
pd.concat([data_one, data_two], ignore_index=True)

A    B    C    D
0   A0   B0  NaN  NaN
1   A1   B1  NaN  NaN
2   A2   B2  NaN  NaN
3   A3   B3  NaN  NaN
4  NaN  NaN   C0   D0
5  NaN  NaN   C1   D1
6  NaN  NaN   C2   D2
7  NaN  NaN   C3   D3

In [111]:
registrations = pd.DataFrame({'reg_id':[1,2,3,4],'name':['Andrew','Bobo','Claire','David']})
logins = pd.DataFrame({'log_id':[1,2,3,4],'name':['Xavier','Andrew','Yolanda','Bobo']})

In [112]:
registrations

reg_id    name
0       1  Andrew
1       2    Bobo
2       3  Claire
3       4   David

In [113]:
logins

log_id     name
0       1   Xavier
1       2   Andrew
2       3  Yolanda
3       4     Bobo

In [114]:
pd.merge(registrations, logins, how='inner',on=['name'])

reg_id    name  log_id
0       1  Andrew       2
1       2    Bobo       4

In [115]:
pd.merge(registrations, logins, how='left',on=['name'])

reg_id    name  log_id
0       1  Andrew     2.0
1       2    Bobo     4.0
2       3  Claire     NaN
3       4   David     NaN

In [116]:
pd.merge(registrations, logins, how='right',on=['name'])

reg_id     name  log_id
0     NaN   Xavier       1
1     1.0   Andrew       2
2     NaN  Yolanda       3
3     2.0     Bobo       4

In [117]:
registrations.columns = ['reg_id', 'reg_name']
registrations

reg_id reg_name
0       1   Andrew
1       2     Bobo
2       3   Claire
3       4    David

In [118]:
pd.merge(registrations, logins, how='outer', left_on='reg_name', right_on='name')

reg_id reg_name  log_id     name
0     1.0   Andrew     2.0   Andrew
1     2.0     Bobo     4.0     Bobo
2     3.0   Claire     NaN      NaN
3     4.0    David     NaN      NaN
4     NaN      NaN     1.0   Xavier
5     NaN      NaN     3.0  Yolanda

In [119]:
messy_names = pd.Series(["andrew  ","bo;bo","  claire  "])
messy_names

0      andrew  
1         bo;bo
2      claire  
dtype: object

In [120]:
messy_names.str.replace(';', '').str.strip().str.capitalize()

0    Andrew
1      Bobo
2    Claire
dtype: object

In [121]:
def cleanup(name):
    name = name.replace(";","")
    name = name.strip()
    name = name.capitalize()
    return name

In [122]:
messy_names.apply(cleanup)

0    Andrew
1      Bobo
2    Claire
dtype: object

In [123]:
import timeit 
  
# code snippet to be executed only once 
setup = '''
import pandas as pd
import numpy as np
messy_names = pd.Series(["andrew  ","bo;bo","  claire  "])
def cleanup(name):
    name = name.replace(";","")
    name = name.strip()
    name = name.capitalize()
    return name
'''
  
# code snippet whose execution time is to be measured 
stmt_pandas_str = ''' 
messy_names.str.replace(";","").str.strip().str.capitalize()
'''

stmt_pandas_apply = '''
messy_names.apply(cleanup)
'''

stmt_pandas_vectorize='''
np.vectorize(cleanup)(messy_names)
'''

In [124]:
timeit.timeit(setup = setup, 
                    stmt = stmt_pandas_str, 
                    number = 10000) 

4.011301329000048

In [125]:
timeit.timeit(setup = setup, 
                    stmt = stmt_pandas_apply, 
                    number = 10000) 

0.7127557290000368

In [126]:
timeit.timeit(setup = setup, 
                    stmt = stmt_pandas_vectorize, 
                    number = 10000) 

0.509286899000017

In [127]:
from datetime import datetime

In [128]:
my_year = 2025
my_month = 1
my_day = 2
my_hour = 13
my_minute = 30
my_second = 15

In [129]:
# January 2nd, 2017
my_date = datetime(my_year,my_month,my_day)
my_date

datetime.datetime(2025, 1, 2, 0, 0)

In [130]:
# January 2nd, 2017 at 13:30:15
my_date_time = datetime(my_year,my_month,my_day,my_hour,my_minute,my_second)
my_date_time

datetime.datetime(2025, 1, 2, 13, 30, 15)

In [131]:
my_date_time.year, my_date_time.month, my_date_time.day

(2025, 1, 2)

In [132]:
myser = pd.Series(['Nov 3, 2024', '2025-01-01', None])

In [133]:
timeser = pd.to_datetime(myser, format='mixed')

In [134]:
timeser[0].year

2024

In [135]:
obvi_euro_date = '31-12-2000'

In [136]:
pd.to_datetime(obvi_euro_date) 

/tmp/ipykernel_8603/272727512.py:1: UserWarning: Parsing dates in %d-%m-%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  pd.to_datetime(obvi_euro_date)


Timestamp('2000-12-31 00:00:00')

In [137]:
euro_date = '10-12-2000'

In [138]:
pd.to_datetime(euro_date) 

Timestamp('2000-10-12 00:00:00')

In [139]:
pd.to_datetime(euro_date,dayfirst=True) 

Timestamp('2000-12-10 00:00:00')

In [140]:
style_date = '12--Dec--2000'

In [141]:
pd.to_datetime(style_date, format='%d--%b--%Y')

Timestamp('2000-12-12 00:00:00')

In [142]:
strange_date = '12th of Dec 2000'
pd.to_datetime(strange_date)

Timestamp('2000-12-12 00:00:00')

In [143]:
strange_date = '12th of Dec 2024'
pd.to_datetime(strange_date)

Timestamp('2024-12-12 00:00:00')

In [144]:
sales = pd.read_csv("https://docs.google.com/spreadsheets/d/1zWW891Nfwmi8uZBYL0Ymk-LuJ-O-LkOVbOeaR1xENaI/export?gid=100964967&format=csv")
sales.head()

DATE  MRTSSM4453USN
0  1992-01-01           1509
1  1992-02-01           1541
2  1992-03-01           1597
3  1992-04-01           1675
4  1992-05-01           1822

In [145]:
type(sales["DATE"][0])

str

In [146]:
sales["DATE"] = pd.to_datetime(sales["DATE"])

In [147]:
sales["DATE"][0].year

1992

In [148]:
sales = pd.read_csv("https://docs.google.com/spreadsheets/d/1zWW891Nfwmi8uZBYL0Ymk-LuJ-O-LkOVbOeaR1xENaI/export?gid=100964967&format=csv",
                    parse_dates=[0])

In [149]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   DATE           340 non-null    datetime64[ns]
 1   MRTSSM4453USN  340 non-null    int64         
dtypes: datetime64[ns](1), int64(1)
memory usage: 5.4 KB


In [150]:
sales = sales.set_index("DATE")

In [151]:
sales.head()

MRTSSM4453USN
DATE                     
1992-01-01           1509
1992-02-01           1541
1992-03-01           1597
1992-04-01           1675
1992-05-01           1822

In [152]:
sales.resample(rule='YE').mean()

MRTSSM4453USN
DATE                     
1992-12-31    1807.250000
1993-12-31    1794.833333
1994-12-31    1841.750000
1995-12-31    1833.916667
1996-12-31    1929.750000
1997-12-31    2006.750000
1998-12-31    2115.166667
1999-12-31    2206.333333
2000-12-31    2375.583333
2001-12-31    2468.416667
2002-12-31    2491.166667
2003-12-31    2539.083333
2004-12-31    2682.416667
2005-12-31    2797.250000
2006-12-31    3001.333333
2007-12-31    3177.333333
2008-12-31    3292.000000
2009-12-31    3353.750000
2010-12-31    3450.083333
2011-12-31    3532.666667
2012-12-31    3697.083333
2013-12-31    3839.666667
2014-12-31    4023.833333
2015-12-31    4212.500000
2016-12-31    4434.416667
2017-12-31    4602.666667
2018-12-31    4830.666667
2019-12-31    4972.750000
2020-12-31    4922.500000

<table border="1" class="colwidths-given docutils">
<colgroup>
<col width="12%" />
<col width="40%" />
<col width="24%" />
<col width="24%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Format Type</th>
<th class="head">Data Description</th>
<th class="head">Reader</th>
<th class="head">Writer</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>text</td>
<td><a class="reference external" href="https://en.wikipedia.org/wiki/Comma-separated_values">CSV</a></td>
<td><a class="reference internal" href="#io-read-csv-table"><span class="std std-ref">read_csv</span></a></td>
<td><a class="reference internal" href="#io-store-in-csv"><span class="std std-ref">to_csv</span></a></td>
</tr>
<tr class="row-odd"><td>text</td>
<td><a class="reference external" href="https://www.json.org/">JSON</a></td>
<td><a class="reference internal" href="#io-json-reader"><span class="std std-ref">read_json</span></a></td>
<td><a class="reference internal" href="#io-json-writer"><span class="std std-ref">to_json</span></a></td>
</tr>
<tr class="row-even"><td>text</td>
<td><a class="reference external" href="https://en.wikipedia.org/wiki/HTML">HTML</a></td>
<td><a class="reference internal" href="#io-read-html"><span class="std std-ref">read_html</span></a></td>
<td><a class="reference internal" href="#io-html"><span class="std std-ref">to_html</span></a></td>
</tr>
<tr class="row-odd"><td>text</td>
<td>Local clipboard</td>
<td><a class="reference internal" href="#io-clipboard"><span class="std std-ref">read_clipboard</span></a></td>
<td><a class="reference internal" href="#io-clipboard"><span class="std std-ref">to_clipboard</span></a></td>
</tr>
<tr class="row-even"><td>binary</td>
<td><a class="reference external" href="https://en.wikipedia.org/wiki/Microsoft_Excel">MS Excel</a></td>
<td><a class="reference internal" href="#io-excel-reader"><span class="std std-ref">read_excel</span></a></td>
<td><a class="reference internal" href="#io-excel-writer"><span class="std std-ref">to_excel</span></a></td>
</tr>
<tr class="row-odd"><td>binary</td>
<td><a class="reference external" href="http://www.opendocumentformat.org">OpenDocument</a></td>
<td><a class="reference internal" href="#io-ods"><span class="std std-ref">read_excel</span></a></td>
<td>&#160;</td>
</tr>
<tr class="row-even"><td>binary</td>
<td><a class="reference external" href="https://support.hdfgroup.org/HDF5/whatishdf5.html">HDF5 Format</a></td>
<td><a class="reference internal" href="#io-hdf5"><span class="std std-ref">read_hdf</span></a></td>
<td><a class="reference internal" href="#io-hdf5"><span class="std std-ref">to_hdf</span></a></td>
</tr>
<tr class="row-odd"><td>binary</td>
<td><a class="reference external" href="https://github.com/wesm/feather">Feather Format</a></td>
<td><a class="reference internal" href="#io-feather"><span class="std std-ref">read_feather</span></a></td>
<td><a class="reference internal" href="#io-feather"><span class="std std-ref">to_feather</span></a></td>
</tr>
<tr class="row-even"><td>binary</td>
<td><a class="reference external" href="https://parquet.apache.org/">Parquet Format</a></td>
<td><a class="reference internal" href="#io-parquet"><span class="std std-ref">read_parquet</span></a></td>
<td><a class="reference internal" href="#io-parquet"><span class="std std-ref">to_parquet</span></a></td>
</tr>
<tr class="row-odd"><td>binary</td>
<td><a class="reference external" href="https://msgpack.org/index.html">Msgpack</a></td>
<td><a class="reference internal" href="#io-msgpack"><span class="std std-ref">read_msgpack</span></a></td>
<td><a class="reference internal" href="#io-msgpack"><span class="std std-ref">to_msgpack</span></a></td>
</tr>
<tr class="row-even"><td>binary</td>
<td><a class="reference external" href="https://en.wikipedia.org/wiki/Stata">Stata</a></td>
<td><a class="reference internal" href="#io-stata-reader"><span class="std std-ref">read_stata</span></a></td>
<td><a class="reference internal" href="#io-stata-writer"><span class="std std-ref">to_stata</span></a></td>
</tr>
<tr class="row-odd"><td>binary</td>
<td><a class="reference external" href="https://en.wikipedia.org/wiki/SAS_(software)">SAS</a></td>
<td><a class="reference internal" href="#io-sas-reader"><span class="std std-ref">read_sas</span></a></td>
<td>&#160;</td>
</tr>
<tr class="row-even"><td>binary</td>
<td><a class="reference external" href="https://docs.python.org/3/library/pickle.html">Python Pickle Format</a></td>
<td><a class="reference internal" href="#io-pickle"><span class="std std-ref">read_pickle</span></a></td>
<td><a class="reference internal" href="#io-pickle"><span class="std std-ref">to_pickle</span></a></td>
</tr>
<tr class="row-odd"><td>SQL</td>
<td><a class="reference external" href="https://en.wikipedia.org/wiki/SQL">SQL</a></td>
<td><a class="reference internal" href="#io-sql"><span class="std std-ref">read_sql</span></a></td>
<td><a class="reference internal" href="#io-sql"><span class="std std-ref">to_sql</span></a></td>
</tr>
<tr class="row-even"><td>SQL</td>
<td><a class="reference external" href="https://en.wikipedia.org/wiki/BigQuery">Google Big Query</a></td>
<td><a class="reference internal" href="#io-bigquery"><span class="std std-ref">read_gbq</span></a></td>
<td><a class="reference internal" href="#io-bigquery"><span class="std std-ref">to_gbq</span></a></td>
</tr>
</tbody>
</table>

In [153]:
import os
os.getcwd()

'/home/alexender/Desktop/Projects/My_Projects/Machine_Learning/Basic/pandas'

In [154]:
df = pd.read_csv("https://docs.google.com/spreadsheets/d/1eHPEC8SErgkYtRBg99AcW2jFNr38AO9DsWIe9ChptRg/export?gid=1147971351&format=csv", 
                 header=None)
df

0   1   2   3
0   a   b   c   d
1   0   1   2   3
2   4   5   6   7
3   8   9  10  11
4  12  13  14  15

In [155]:
df = pd.read_csv("https://docs.google.com/spreadsheets/d/1eHPEC8SErgkYtRBg99AcW2jFNr38AO9DsWIe9ChptRg/export?gid=1147971351&format=csv", 
                 # index_col=0
                )
df

a   b   c   d
0   0   1   2   3
1   4   5   6   7
2   8   9  10  11
3  12  13  14  15

In [156]:
df.to_csv('/home/alexender/Desktop/Projects/My_Projects/Data/example_file_if_want_delete.csv', index=False)

In [157]:
new = pd.read_csv('/home/alexender/Desktop/Projects/My_Projects/Data/example_file_if_want_delete.csv')
new

a   b   c   d
0   0   1   2   3
1   4   5   6   7
2   8   9  10  11
3  12  13  14  15

In [158]:
!pip install lxml

Looking in indexes: https://pypi.org/simple, https://pypi.ngc.nvidia.com


In [159]:
tables = pd.read_html('https://en.wikipedia.org/wiki/World_population')

In [160]:
population_percentage = tables[4]

In [161]:
len(tables)

30

In [162]:
population_percentage

Country / Dependency  Population % of world         Date  \
0                India  1425775850      17.5%  14 Apr 2023   
1                China  1409670000      17.3%  17 Jan 2024   
2        United States   337441162      4.14%  25 Jan 2025   
3            Indonesia   278696200      3.42%   1 Jul 2023   
4             Pakistan   229488994      2.81%   1 Jul 2022   
5              Nigeria   216746934      2.66%   1 Jul 2022   
6               Brazil   218714895      2.68%  25 Jan 2025   
7           Bangladesh   168220000      2.06%   1 Jul 2020   
8               Russia   147190000      1.80%   1 Oct 2021   
9               Mexico   128271248      1.57%  31 Mar 2022   

  Source (official or from the United Nations)  
0                            UN projection[89]  
1                 National annual estimate[90]  
2                National population clock[91]  
3                 National annual estimate[92]  
4                            UN projection[93]  
5                            UN projection[93]  
6                National population clock[94]  
7               Annual Population Estimate[95]  
8          2021 preliminary census results[96]  
9                                          NaN

In [163]:
population_percentage.to_html('/home/alexender/Desktop/Projects/My_Projects/Data/sample_if_want_delete.html', index=False)

In [164]:
!pip install openpyxl
!pip install xlrd

Looking in indexes: https://pypi.org/simple, https://pypi.ngc.nvidia.com
Looking in indexes: https://pypi.org/simple, https://pypi.ngc.nvidia.com


In [165]:
df = pd.read_excel("https://docs.google.com/spreadsheets/d/15R4_3PBRo6mJhrgkaIRNQYByjpQWDAdV/export?gid=1868581063&format=xlsx", 
                 sheet_name='First_Sheet')
df

a   b   c   d
0   0   1   2   3
1   4   5   6   7
2   8   9  10  11
3  12  13  14  15

In [166]:
wb = pd.ExcelFile('https://docs.google.com/spreadsheets/d/15R4_3PBRo6mJhrgkaIRNQYByjpQWDAdV/export?gid=1868581063&format=xlsx')

In [167]:
wb.sheet_names

['First_Sheet']

In [168]:
excel_sheet_dict = pd.read_excel("https://docs.google.com/spreadsheets/d/15R4_3PBRo6mJhrgkaIRNQYByjpQWDAdV/export?gid=1868581063&format=xlsx", 
                 sheet_name=None)

In [169]:
our_df = excel_sheet_dict['First_Sheet']

In [170]:
our_df.to_excel('/home/alexender/Desktop/Projects/My_Projects/Data/excel_if_want_delete.xlsx', sheet_name='First_Sheet', index=False)

In [171]:
!pip install SQLAlchemy

Looking in indexes: https://pypi.org/simple, https://pypi.ngc.nvidia.com


In [181]:
from sqlalchemy import create_engine
import urllib
 
# DEFINE THE DATABASE CREDENTIALS
user = 'Alex'
password = 'Alex@$14798|<</>>'
host = 'localhost'
port = 3306
database = 'Personal_DB'
decoded_password = urllib.parse.quote_plus(password)

In [182]:
engine = create_engine(
        url="mysql+pymysql://{0}:{1}@{2}:{3}/{4}".format(
            user, decoded_password, host, port, database
        )
    )
sql_query = """SELECT * FROM Personal_DB.Telco_customer_churn;"""

In [183]:
sql_df = pd.read_sql(sql=sql_query, con=engine)

In [184]:
sql_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7032 entries, 0 to 7031
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         7032 non-null   object 
 1   Count              7032 non-null   object 
 2   Country            7032 non-null   object 
 3   State              7032 non-null   object 
 4   City               7032 non-null   object 
 5   Zip Code           7032 non-null   int64  
 6   Lat Long           7032 non-null   object 
 7   Latitude           7032 non-null   float64
 8   Longitude          7032 non-null   float64
 9   Gender             7032 non-null   object 
 10  Senior Citizen     7032 non-null   object 
 11  Partner            7032 non-null   object 
 12  Dependents         7032 non-null   object 
 13  Tenure Months      7032 non-null   object 
 14  Phone Service      7032 non-null   object 
 15  Multiple Lines     7032 non-null   object 
 16  Internet Service   7032 

In [185]:
sql_df.isna().sum()

CustomerID           0
Count                0
Country              0
State                0
City                 0
Zip Code             0
Lat Long             0
Latitude             0
Longitude            0
Gender               0
Senior Citizen       0
Partner              0
Dependents           0
Tenure Months        0
Phone Service        0
Multiple Lines       0
Internet Service     0
Online Security      0
Online Backup        0
Device Protection    0
Tech Support         0
Streaming TV         0
Streaming Movies     0
Contract             0
Paperless Billing    0
Payment Method       0
Monthly Charges      0
Total Charges        0
Churn Label          0
Churn Value          0
Churn Score          0
CLTV                 0
Churn Reason         0
dtype: int64

In [186]:
temp_db = create_engine("sqlite:///:memory:") # to create the temp DB in our RAM

In [187]:
temp_db

Engine(sqlite:///:memory:)

In [188]:
temp_df_db = pd.DataFrame(data=np.random.randint(low=10, high=100, size=(4,4)), columns=['a', 'b', 'c', 'd'])
temp_df_db

a   b   c   d
0  35  88  84  85
1  43  19  53  30
2  92  40  13  56
3  39  57  37  91

In [189]:
temp_df_db.to_sql(name='new_table', con=temp_db, if_exists='append', index=False)

4

In [190]:
pd.read_sql(sql='new_table', con=temp_db)

a   b   c   d
0  35  88  84  85
1  43  19  53  30
2  92  40  13  56
3  39  57  37  91

In [191]:
pd.read_sql_query(sql="SELECT a,c FROM new_table",con=temp_db)

a   c
0  35  84
1  43  53
2  92  13
3  39  37

In [192]:
df = pd.read_csv("https://docs.google.com/spreadsheets/d/1Agin0trPHPYIT_tFcuVUjbpeQ2qsLlIeaFhVkt_HgdE/export?gid=1240789564&format=csv")
df

Account Number       Company       Contact Account Manager  \
0          2123398        Google   Larry Pager    Edward Thorp   
1          2123398        Google   Larry Pager    Edward Thorp   
2          2123398        Google   Larry Pager    Edward Thorp   
3          2192650          BOBO   Larry Pager    Edward Thorp   
4           420496          IKEA     Elon Tusk    Edward Thorp   
5           636685    Tesla Inc.     Elon Tusk    Edward Thorp   
6           636685    Tesla Inc.     Elon Tusk    Edward Thorp   
7          1216870     Microsoft   Will Grates    Edward Thorp   
8          2200450       Walmart   Will Grates    Edward Thorp   
9           405886         Apple  Cindy Phoner  Claude Shannon   
10          470248  Exxon Mobile  Cindy Phoner  Claude Shannon   
11          698032           ATT  Cindy Phoner  Claude Shannon   
12          698032           ATT  Cindy Phoner  Claude Shannon   
13          902797    CVS Health  Emma Gordian  Claude Shannon   
14         2046943    Salesforce  Emma Gordian  Claude Shannon   
15         2169499         Cisco  Emma Gordian  Claude Shannon   
16         2169499         Cisco  Emma Gordian  Claude Shannon   

            Product  Licenses  Sale Price        Status  
0         Analytics       150     2100000     Presented  
1        Prediction       150      700000     Presented  
2          Tracking       300      350000  Under Review  
3         Analytics       150     2450000          Lost  
4         Analytics       300     4550000           Won  
5         Analytics       300     2800000  Under Review  
6        Prediction       150      700000     Presented  
7          Tracking       300      350000  Under Review  
8         Analytics       150     2450000          Lost  
9         Analytics       300     4550000           Won  
10        Analytics       150     2100000     Presented  
11         Tracking       150      350000  Under Review  
12       Prediction       150      700000     Presented  
13         Tracking       450      490000           Won  
14        Analytics       750     7000000           Won  
15        Analytics       300     4550000          Lost  
16  GPS Positioning       300      350000     Presented

In [194]:
licenses = df[['Company', 'Product', 'Licenses']]

In [198]:
pd.pivot(data=licenses, index='Company', columns='Product')

Licenses                                    
Product      Analytics GPS Positioning Prediction Tracking
Company                                                   
ATT                NaN             NaN      150.0    150.0
Apple            300.0             NaN        NaN      NaN
BOBO             150.0             NaN        NaN      NaN
CVS Health         NaN             NaN        NaN    450.0
Cisco            300.0           300.0        NaN      NaN
Exxon Mobile     150.0             NaN        NaN      NaN
Google           150.0             NaN      150.0    300.0
IKEA             300.0             NaN        NaN      NaN
Microsoft          NaN             NaN        NaN    300.0
Salesforce       750.0             NaN        NaN      NaN
Tesla Inc.       300.0             NaN      150.0      NaN
Walmart          150.0             NaN        NaN      NaN

In [208]:
pd.pivot_table(df, index='Company', aggfunc='sum', values=['Licenses', 'Sale Price'])

Licenses  Sale Price
Company                           
ATT                300     1050000
Apple              300     4550000
BOBO               150     2450000
CVS Health         450      490000
Cisco              600     4900000
Exxon Mobile       150     2100000
Google             600     3150000
IKEA               300     4550000
Microsoft          300      350000
Salesforce         750     7000000
Tesla Inc.         450     3500000
Walmart            150     2450000

In [205]:
df.groupby("Company").sum(numeric_only=True)

Account Number  Licenses  Sale Price
Company                                           
ATT                  1396064       300     1050000
Apple                 405886       300     4550000
BOBO                 2192650       150     2450000
CVS Health            902797       450      490000
Cisco                4338998       600     4900000
Exxon Mobile          470248       150     2100000
Google               6370194       600     3150000
IKEA                  420496       300     4550000
Microsoft            1216870       300      350000
Salesforce           2046943       750     7000000
Tesla Inc.           1273370       450     3500000
Walmart              2200450       150     2450000

In [212]:
pd.pivot_table(df, index = ['Account Manager', 'Contact'], values=['Sale Price'], columns=['Product'], aggfunc='sum', fill_value=0)

Sale Price                                    
Product                       Analytics GPS Positioning Prediction Tracking
Account Manager Contact                                                    
Claude Shannon  Cindy Phoner    6650000               0     700000   350000
                Emma Gordian   11550000          350000          0   490000
Edward Thorp    Elon Tusk       7350000               0     700000        0
                Larry Pager     4550000               0     700000   350000
                Will Grates     2450000               0          0   350000

In [213]:
pd.pivot_table(df, index = ['Account Manager', 'Contact', 'Product'], values=['Sale Price'], aggfunc='sum', fill_value=0)

Sale Price
Account Manager Contact      Product                    
Claude Shannon  Cindy Phoner Analytics           6650000
                             Prediction           700000
                             Tracking             350000
                Emma Gordian Analytics          11550000
                             GPS Positioning      350000
                             Tracking             490000
Edward Thorp    Elon Tusk    Analytics           7350000
                             Prediction           700000
                Larry Pager  Analytics           4550000
                             Prediction           700000
                             Tracking             350000
                Will Grates  Analytics           2450000
                             Tracking             350000

In [214]:
pd.pivot_table(df, index = ['Account Manager', 'Contact', 'Product'], values=['Sale Price'], aggfunc='sum', fill_value=0, margins=True)

Sale Price
Account Manager Contact      Product                    
Claude Shannon  Cindy Phoner Analytics           6650000
                             Prediction           700000
                             Tracking             350000
                Emma Gordian Analytics          11550000
                             GPS Positioning      350000
                             Tracking             490000
Edward Thorp    Elon Tusk    Analytics           7350000
                             Prediction           700000
                Larry Pager  Analytics           4550000
                             Prediction           700000
                             Tracking             350000
                Will Grates  Analytics           2450000
                             Tracking             350000
All                                             36540000